KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,388 @@
+        end
+        if !isempty(table_gradients)
+            @printf(fid, "# Format of arbitrary gradients:\n")
+            @printf(
+                fid,
+                "#   time_shape_id of 0 means default timing (stepping with grad_raster starting at 1/2 of grad_raster)\n"
+            )
+            @printf(fid, "# id amplitude amp_shape_id time_shape_id delay\n") # do we need delay ???
+            @printf(fid, "# ..      Hz/m       ..         ..          us\n")
+            @printf(fid, "[GRADIENTS]\n")
+            for (id, _, amp, amp_id, time_id, delay) in table_gradients
+                @printf(fid, "%d %12g %d %d %d\n", id, amp, amp_id, time_id, delay)
+            end
+            @printf(fid, "\n")
+        end
+        if !isempty(table_trap)
+            @printf(fid, "# Format of trapezoid gradients:\n")
+            @printf(fid, "# id amplitude rise flat fall delay\n")
+            @printf(fid, "# ..      Hz/m   us   us   us    us\n")
+            @printf(fid, "[TRAP]\n")
+            for (id, _, amp, rise, flat, fall, delay) in table_trap
+                @printf(fid, "%2d %12g %3d %4d %3d %3d\n", id, amp, rise, flat, fall, delay)
+            end
+            @printf(fid, "\n")
+        end
+        if !isempty(table_adc)
+            @printf(fid, "# Format of ADC events:\n")
+            @printf(fid, "# id num dwell delay freq phase\n")
+            @printf(fid, "# ..  ..    ns    us   Hz   rad\n")
+            @printf(fid, "[ADC]\n")
+            for (id, _, num, dwell, delay, freq, phase) in table_adc
+                @printf(fid, "%d %d %.0f %.0f %g %g\n", id, num, dwell, delay, freq, phase)
+            end
+            @printf(fid, "\n")
+        end
+        if !isempty(shape_data_id_num)
+            @printf(fid, "# Sequence Shapes\n")
+            @printf(fid, "[SHAPES]\n\n")
+            for (data, id, num) in shape_data_id_num
+                @printf(fid, "shape_id %d\n", id)
+                @printf(fid, "num_samples %d\n", num)
+                [@printf(fid, "%.9g\n", datai) for datai in data]
+                @printf(fid, "\n")
+            end
+        end
+    end
+    md5hash = bytes2hex(open(md5, filename))
+    open(filename, "a") do fid
+        @printf(fid, "\n[SIGNATURE]\n") # the preceding new line BELONGS to the signature (and needs to be sripped away to recalculate the signature)
+        @printf(
+            fid,
+            "# This is the hash of the Pulseq file, calculated right before the [SIGNATURE] section was added\n"
+        )
+        @printf(
+            fid,
+            "# It can be reproduced/verified with md5sum if the file trimmed to the position right above [SIGNATURE]\n"
+        )
+        @printf(
+            fid,
+            "# The new line character preceding [SIGNATURE] BELONGS to the signature (and needs to be sripped away for recalculating/verification)\n"
+        )
+        @printf(fid, "Type md5\n")
+        @printf(fid, "Hash %s\n", md5hash)
```

Comment:
```suggestion
    md5hash = bytes2hex(open(md5, filename))
    open(filename, "a") do fid
        @printf(
            fid,
            """
            [SIGNATURE]
            # This is the hash of the Pulseq file, calculated right before the [SIGNATURE] section was added
            # It can be reproduced/verified with md5sum if the file trimmed to the position right above [SIGNATURE]
            # The new line character preceding [SIGNATURE] BELONGS to the signature (and needs to be sripped away for recalculating/verification)
            Type md5
            Hash %s
            """,
            md5hash
        )
```